edb/server/compiler/explain/__init__.py

```diff
@@ -0,0 +1,131 @@
+#
+# This source file is part of the EdgeDB open source project.
+#
+# Copyright 2016-present MagicStack Inc. and the EdgeDB authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+from __future__ import annotations
+from typing import *
+
+import json
+import logging
+import pickle
+import dataclasses
+
+from edb.common import debug
+from edb.edgeql import ast as qlast
+from edb.ir import ast as irast
+from edb.schema import schema as s_schema
+from edb.pgsql import ast as pgast
+
+from . import coarse_grained
+from . import fine_grained
+from . import ir_analyze
+from . import pg_tree
+from . import to_json
+
+
+log = logging.getLogger(__name__)
+
+
+# "affects_compilation" config vals that we don't actually want to report out.
+# This turns out to be a majority of them
+OMITTED_CONFIG_VALS = {
+    "allow_dml_in_functions", "allow_bare_ddl", "force_database_error",
+}
+
+
+@dataclasses.dataclass
+class Arguments(to_json.ToJson):
+    execute: bool
+    buffers: bool
+
+
+def analyze_explain_output(
+    query_asts_pickled: bytes,
+    data: list[list[bytes]],
+    std_schema: s_schema.FlatSchema,
+) -> bytes:
+    if debug.flags.edgeql_explain:
+        debug.header('Explain')
+
+    ql: qlast.Base
+    ir: irast.Statement
+    pg: pgast.Base
+    ql, ir, pg, (config_vals, args) = pickle.loads(query_asts_pickled)
+    args = Arguments(**args)
+
+    schema = ir.schema
+    # We omit the std schema when serializing, so put it back
+    if isinstance(schema, s_schema.ChainedSchema):
+        schema = s_schema.ChainedSchema(
+            top_schema=schema._top_schema,
+            global_schema=schema._global_schema,
+            base_schema=std_schema
+        )
+
+    assert len(data) == 1 and len(data[0]) == 1
+    plan = json.loads(data[0][0])
+    assert len(plan) == 1
+    plan = debug_tree = plan[0]['Plan']
+
+    if debug.flags.edgeql_explain:
+        import importlib
+        importlib.reload(pg_tree)
+        importlib.reload(ir_analyze)
+        importlib.reload(fine_grained)
+        importlib.reload(coarse_grained)
```

Comment:
Elvis and I were just chatting and I think we should add a quick way to kill all the compiler worker processes, which will lead to them getting reloaded without needing to restart postgres or the edgedb server.
We were thinking maybe a flag that makes it so that a single ctrl-c just cycles the workers and a double ctrl-c actually stops the server.

Reply:
There is also common Ctrl+\ which sends SIGQUIT. This can be used instead of making Ctrl+C confusing.